Global exception handler and data validation.

ponomvrenko opened the pull request.

ponomvrenko committed:
Set up a proper naming for update by id method. Also, it starts throw exception inside updateById method in service layer.

ponomvrenko committed:
Configured in pom.xml dependency: hibernate-validator. Also added meaningful validation for existing Dto. Created CustomGlobalExceptionHandler.

ponomvrenko committed:
Added handleEntityNotFoundException in CustomGlobalExceptionHandler.

fedorovychh commented on pom.xml:
```diff
@@ -80,6 +80,11 @@
             <artifactId>liquibase-maven-plugin</artifactId>
             <version>${liquibase.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-validator</artifactId>
+            <version>8.0.1.Final</version>
+        </dependency>
```
can be changed to 
```
 <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-validation</artifactId>
  </dependency>
```
because of Spring project

vovasalyha replied:
Not resolved.

ponomvrenko replied:
@fedorovychh thanks for code example!

fedorovychh commented on src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java:
```diff
@@ -1,13 +1,20 @@
 
 @Data
 public class CreateBookRequestDto {
+    @NotNull
     private String title;
```
You need to validate length here

vovasalyha replied:
Not resolved.

fedorovychh commented on src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java:
```diff
@@ -1,13 +1,20 @@
     private String title;
+    @NotNull
     private String author;
+    @NotNull
     private String isbn;
```
ISBN consists from digits and it has [fixed length](https://uk.wikipedia.org/wiki/ISBN)

vovasalyha replied:
Not resolved.

vovasalyha commented on src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java:
```diff
@@ -45,9 +45,13 @@ public void deleteById(Long id) {
+            book.setId(id);
+            return bookMapper.toDto(bookRepository.save(book));
+        } catch (EntityNotFoundException e) {
+            throw new EntityNotFoundException("Can't update book by id: " + id);
```
`Can't update book by id:` is too generic message. `Book by ID {id} not found` would be more meaningful.

vovasalyha reviewed (changes requested):
Looks good, but see a couple of comments.

ponomvrenko committed:
Configuration:
- pom.xml: replaced dependency <hibernate-valid> by more proper <spring-boot-starter-validation>.

DTO:
- CreateBookRequestDto: added @Size to String fields with proper constraints. (ISBN, title)

Service:
- BookServiceImpl: made exception messages human-friendly

bhdnchui reviewed (approved):
Good job!
